src/hooks/useGridDimensions.ts

```diff
@@ -1,46 +1,48 @@
-import { useLayoutEffect, useRef, useState } from 'react';
-import { flushSync } from 'react-dom';
+import { useCallback, useRef, useSyncExternalStore } from 'react';
 
-export function useGridDimensions() {
-  const gridRef = useRef<HTMLDivElement>(null);
-  const [inlineSize, setInlineSize] = useState(1);
-  const [blockSize, setBlockSize] = useState(1);
-  const [horizontalScrollbarHeight, setHorizontalScrollbarHeight] = useState(0);
+type Dimensions = [inlineSize: number, blockSize: number, horizontalScrollbarHeight: number];
 
-  useLayoutEffect(() => {
-    const { ResizeObserver } = window;
+const defaultDimensions: Dimensions = [1, 1, 0];
 
-    // don't break in Node.js (SSR), jsdom, and browsers that don't support ResizeObserver
-    // eslint-disable-next-line @typescript-eslint/no-unnecessary-condition
-    if (ResizeObserver == null) return;
+function getServerSnapshot() {
+  return defaultDimensions;
+}
+
+export function useGridDimensions() {
+  const gridRef = useRef<HTMLDivElement>(null);
+  const dimensionsRef = useRef(defaultDimensions);
 
-    const { clientWidth, clientHeight, offsetWidth, offsetHeight } = gridRef.current!;
+  const subscribe = useCallback((callback: () => void) => {
+    const element = gridRef.current!;
+    const { clientWidth, clientHeight, offsetWidth, offsetHeight } = element;
     const { width, height } = gridRef.current!.getBoundingClientRect();
     const initialHorizontalScrollbarHeight = offsetHeight - clientHeight;
     const initialWidth = width - offsetWidth + clientWidth;
     const initialHeight = height - initialHorizontalScrollbarHeight;
-
-    setInlineSize(initialWidth);
-    setBlockSize(initialHeight);
-    setHorizontalScrollbarHeight(initialHorizontalScrollbarHeight);
+    dimensionsRef.current = [initialWidth, initialHeight, initialHorizontalScrollbarHeight];
 
     const resizeObserver = new ResizeObserver((entries) => {
       const size = entries[0].contentBoxSize[0];
-      const { clientHeight, offsetHeight } = gridRef.current!;
-
-      // we use flushSync here to avoid flashing scrollbars
-      flushSync(() => {
-        setInlineSize(size.inlineSize);
-        setBlockSize(size.blockSize);
-        setHorizontalScrollbarHeight(offsetHeight - clientHeight);
-      });
+      const { clientHeight, offsetHeight } = element;
+      dimensionsRef.current = [size.inlineSize, size.blockSize, offsetHeight - clientHeight];
+      callback();
     });
-    resizeObserver.observe(gridRef.current!);
+    resizeObserver.observe(element);
 
     return () => {
       resizeObserver.disconnect();
     };
   }, []);
 
+  const getSnapShot = useCallback(() => {
+    return dimensionsRef.current;
```

Comment:
Don't we read `.current` during render since we call `getSnapShot()` to intialize the `useSyncExternalStore`?